modules/tlsinfo.cpp

```diff
@@ -0,0 +1,329 @@
+		AddCommand("Send-Data", static_cast<CModCommand::ModCmdFunc>(&CTlsInfoMod::SendCertificateCommand), "", "Send certificate information for the active connection in an easily parsable format for application developers.");
+	}
+
+	void OnClientCapLs(CClient *mClient, SCString &mCaps) override
+	{
+		mCaps.insert(TlsInfoCap);
+	}
+
+	bool IsClientCapSupported(CClient *mClient, const CString &mCap, bool mState) override
+	{
+		return mCap.Equals(TlsInfoCap);
+	}
+
+	void PrintCipherCommand(const CString &mLine)
+	{
+		CClient *mClient;
+
+		SSL *mSSLObject;
+
+		if (GetRelevantObjects(nullptr, &mClient, nullptr, &mSSLObject) == false) {
+			return;
+		}
+
+		/* Determine the protocol used for the connection */
+		int mSSLVersion = SSL_version(mSSLObject);
+
+		CString mSSLVersionString;
+
+		if (mSSLVersion == TLS1_2_VERSION) {
+			mSSLVersionString = CString("Transport Layer Security (TLS), version 1.2");
+		} else if (mSSLVersion == TLS1_1_VERSION) {
+			mSSLVersionString = CString("Transport Layer Security (TLS), version 1.1");
+		} else if (mSSLVersion == TLS1_VERSION) {
+			mSSLVersionString = CString("Transport Layer Security (TLS), version 1.0");
+		} else if (mSSLVersion == SSL3_VERSION) {
+			mSSLVersionString = CString("Secure Sockets Layer (SSL), version 3.0");
+		} else {
+			mSSLVersionString = CString("Unknown");
+		}
+
+		/* Output cipher information */
+		CString mCipherNameString = CString(SSL_get_cipher_name(mSSLObject));
+
+		PutModule("Connection secured using " + mSSLVersionString + " with the cipher suite: " + mCipherNameString);
+	}
+
+	void PrintCertificateCommand(const CString &mLine)
+	{
+		bool mPrintCertificateParents = false;
+
+		CString sCmd = mLine.Token(1);
+
+		if (sCmd.Equals("details")) {
+			mPrintCertificateParents = true;
+		}
+
+		PresentCertificateInformation(mLine, true, mPrintCertificateParents);
+	}
+
+	void SendCertificateCommand(const CString &mLine)
+	{
+		PresentCertificateInformation(mLine, false, true);
+	}
+
+	void PresentCertificateInformation(const CString &mLine, bool mPrintCertificates = false, bool mPrintCertificateParents = false)
+	{
+		CClient *mClient;
+
+		SSL *mSSLObject;
+
+		if (GetRelevantObjects(nullptr, &mClient, nullptr, &mSSLObject) == false) {
+			return;
+		}
+
+		if (mClient->IsCapEnabled(TlsInfoCap) == false ||
+			mClient->HasBatch() == false)
+		{
+			mPrintCertificates = true;
+		}
+
+		/* Obtain list of certificates from SSL context object */
+		STACK_OF(X509) *mCertificateChain = SSL_get_peer_cert_chain(mSSLObject);
+
+		if (mCertificateChain == NULL) {
+			PutModule("Error: SSL_get_peer_cert_chain() returned nullptr");
+
+			return;
+		}
+
+		if (mPrintCertificates) {
+			PrintCertificateChainToQuery(mClient, mCertificateChain, mPrintCertificateParents);
+		} else {
+			SendCertificateChain(mClient, mCertificateChain);
+		}
+	}
+
+private:
+	bool GetRelevantObjects(CIRCNetwork **eNetwork, CClient **eClient, CIRCSock **eSocket, SSL **eSSLObject)
```

Comment:
What's the benefit of this function over `GetClient()` etc?
In the only invocation of it half of parameters are not even used.